File: 1991-find-the-middle-index-in-array/1991-find-the-middle-index-in-array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_middle_index(nums)
    total = nums.sum
    left_sum = 0
    nums.each_with_index do |current_value, index|
        right_sum = total - left_sum - current_value
        return index if left_sum == right_sum
        left_sum += current_value
    end
    -1
end

File: 1991-find-the-middle-index-in-array/1991-find-the-middle-index-in-array.cpp
int init=[]{ios_base::sync_with_stdio(false);cin.tie(0);cout.tie(0);return 0;}();
class Solution {
public:
    int findMiddleIndex(vector<int>& nums) {
        int total = accumulate(nums.begin(), nums.end(), 0);
        int left_sum = 0, right_sum = 0;

        for (int index=0;index<nums.size();++index) {
            int cur_value = nums[index];
            right_sum = total - left_sum - cur_value;
            if (left_sum == right_sum) return index;
            left_sum += cur_value;
        }
        return -1;
    }
};

